src/node_perf.cc

```diff
@@ -0,0 +1,508 @@
+    CHECK_EQ(entry, env->get_performance_timeline_head<PerformanceEntry>());
+    PerformanceEntry* next = entry->next_;
+    if (next != nullptr) {
+      CHECK_EQ(entry, next->previous_);
+      entry->next_ = nullptr;
+      next->previous_ = nullptr;
+      env->set_performance_timeline_head(next);
+    } else {
+      // this is the only item in the timeline! Reset the timeline
+      env->set_performance_timeline_head(nullptr);
+      env->set_performance_timeline_tail(nullptr);
+    }
+  } else {
+    // this is in the middle or the end of the timeline
+    PerformanceEntry* previous = entry->previous_;
+    PerformanceEntry* next = entry->next_;
+    previous->next_ = next;
+    if (next == nullptr) {
+      CHECK_EQ(entry, env->get_performance_timeline_tail<PerformanceEntry>());
+      // entry should be the tail, so make previous the new tail
+      env->set_performance_timeline_tail(previous);
+    } else {
+      next->previous_ = previous;
+    }
+  }
+  entry->removed_ = true;
+}
+
+inline NodePerformanceEntryTypes GetPerformanceEntryType(const char* type) {
+#define V(name, label)                                                        \
+  do {                                                                        \
+    if (strcmp(label, type) == 0)                                             \
+      return NODE_PERFORMANCE_ENTRY_TYPE_##name;                              \
+  } while (0);
+  NODE_PERFORMANCE_ENTRY_TYPES(V)
+#undef V
+  return NODE_PERFORMANCE_ENTRY_TYPE_COUNT;
+}
+
+inline void NotifyObservers(Environment* env, PerformanceEntry* entry) {
+  int32_t* observers = env->get_performance_observers();
+  NodePerformanceEntryTypes type =
+      GetPerformanceEntryType(entry->type().c_str());
+  if (observers == nullptr ||
+      type == NODE_PERFORMANCE_ENTRY_TYPE_COUNT ||
+      !observers[type]) {
+    return;
+  }
+  Local<Context> context = env->context();
+  Isolate* isolate = env->isolate();
+  Local<Value> argv = entry->object();
+  env->performance_entry_callback()->Call(context,
+                                          v8::Undefined(isolate),
+                                          1, &argv);
+}
+
+void PerformanceEntry::AddToTimeline(Environment* env,
+                                     PerformanceEntry* entry) {
+  auto tail = env->get_performance_timeline_tail<PerformanceEntry>();
+  if (tail == nullptr) {
+    // if there is no tail, the timeline is new...
+    env->set_performance_timeline_head(entry);
+    env->set_performance_timeline_tail(entry);
+    goto notify;
+  }
+  // Otherwise attach to the end
+  entry->previous_ = tail;
+  tail->next_ = entry;
+  env->set_performance_timeline_tail(entry);
+ notify:
+  NotifyObservers(env, entry);
+}
+
+void PerformanceFunction::New(const FunctionCallbackInfo<Value>& args) {}
+
+void PerformanceGC::New(const FunctionCallbackInfo<Value>& args) {}
+
+void PerformanceMark::New(const FunctionCallbackInfo<Value>& args) {
```

Comment:
FWIW the browser's `PerformanceMark` class constructor always throws an error when called. The fact that we are not exporting this class (it's available in browsers as a global) means it's probably fine, just something to be aware of.